src/shiv/bootstrap/__init__.py

```diff
@@ -73,15 +73,28 @@ def extract_site_packages(archive, target_path, compile_pyc, compile_workers=0):
     """
     target_path_tmp = Path(target_path.parent, target_path.stem + ".tmp")
 
-    for filename in archive.namelist():
-        if filename.startswith("site-packages"):
-            archive.extract(filename, target_path_tmp)
+    if target_path_tmp.exists():
 
-    if compile_pyc:
-        compileall.compile_dir(target_path_tmp, quiet=2, workers=compile_workers)
+        # it's possible that another instance of the invoked pyz is currently being extracted, so
+        # we'll wait until it's done to ensure atomicity - if a previous invocations target_path_tmp
+        # path is orphaned, this could potentially stall forever :(
+        while target_path_tmp.exists() and not target_path.exists():
```

Comment:
how to prevent this from potentially stalling forever

without resorting to have to vendor in a cross platform file locking library :(